src/Modules/UngradedModules/ModuleGens.jl

```diff
@@ -401,31 +401,65 @@ Note: `:via_lift` is typically faster than `:via_transform` for a single vector
 is faster if many vectors are lifted
 """
 function coordinates(a::FreeModElem, M::SubModuleOfFreeModule, task::Symbol = :auto)
-  if iszero(a)
-    return sparse_row(base_ring(parent(a)))
-  end
-  if task == :auto
-    if coefficient_ring(base_ring(parent(a))) isa Field #base_ring(base_ring(...)) does not work for MPolyQuos
-      task = :via_transform
+    if iszero(a)
+        return sparse_row(base_ring(parent(a)))
+    end
+    return coordinates_atomic(a, M; task=task)
+end
+
+function coordinates_atomic(a::FreeModElem{T}, M::SubModuleOfFreeModule; task::Symbol = :auto) where {S<:Union{ZZRingElem,<:FieldElem}, T<:MPolyRingElem{S}}
+    if task == :auto
+        task = coefficient_ring(base_ring(parent(a))) isa Field ? :via_transform : :via_lift
+    end
+    for i in 1:ngens(M)
+        g = gen(M,i)
+        if a == g
+            R = base_ring(M)
+            return sparse_row(R, [(i,R(1))])
+        end
+    end
+    if task == :via_transform
+        std, _ = lift_std(M)
+        return coordinates_via_transform(a, std)
+    elseif task == :via_lift
+        return coordinates(a, M.gens)
     else
-      task = :via_lift
+        error("Invalid task given.")
     end
-  end
-  for i in 1:ngens(M)
-    g = gen(M,i)
-    if a == g
-      R = base_ring(M)
-      return sparse_row(R, [(i,R(1))])
+end
+
+function sparse_row_from_dense(R::Ring, dense_vec::Vector{T}) where T
```

Comment:
This function already exists as `sparse_row(::MatElem)` (for Oscar matrices instead of julia vectors), but from the first glance it looks like as if in all places you call this you actually have an Oscar matrix and convert it to a vector first. This could be avoided and you then don't need this helper function.

If you find out that you cannot get rid of it, please open a PR with this to Hecke before having this PR here merged. (Moving it afterwards is a lot more work than doing it before merging)